### PR TITLE
Add gpu sharing in predicate logic.

### PR DESCRIPTION
**Pull request by tizhou86**
Add gpu sharing in predicate logic, will add gpu sharing in allocate logic in next pr.

### CODE REVIEW

File: pkg/scheduler/api/device_info.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func (di *DeviceInfo) GetUsedGPUMemory() uint {
+	res := uint(0)
+	for _, pod := range di.PodMap {
+		if pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed {
+			continue
+		} else {
+			if len(pod.ObjectMeta.Annotations) > 0 {
+				mem, found := pod.ObjectMeta.Annotations["volcano.sh/pod-gpu-memory"]
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/device_info.go:54]: declare a const instead

File: pkg/scheduler/api/node_info.go [hunk 2]
```diff
@@ -314,3 +346,60 @@ func (ni *NodeInfo) Pods() (pods []*v1.Pod) {
...
 	return
 }
+
+func (ni *NodeInfo) CheckPredicatePodOnGPUNode(pod *v1.Pod) bool {
+	res := false
+	memReq := uint(0)
+
+	remainMems := ni.GetDevicesRemainGPUMemory()
+	if len(pod.ObjectMeta.Annotations) > 0 {
+		mem, found := pod.ObjectMeta.Annotations["volcano.sh/pod-gpu-memory"]
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/node_info.go:356]: Why not use resource request?

File: pkg/scheduler/api/node_info.go [hunk 3]
```diff
@@ -314,3 +346,60 @@ func (ni *NodeInfo) Pods() (pods []*v1.Pod) {
...
+			m, _ := strconv.Atoi(mem)
+			memReq = uint(m)
+		}
+	}
+
+	if len(remainMems) > 0 {
+		for devID := 0; devID < len(ni.Devices); devID++ {
+			availableGPU, ok := remainMems[devID]
+			if ok {
+				if availableGPU >= memReq {
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/node_info.go:367]: I understand the devices here is split by gpu core from `SetNodeGPUInfo` above.
And we can only allocate one device for a pod. My question is why can't we allocate more than one device?

File: pkg/scheduler/api/node_info.go [hunk 4]
```diff
@@ -154,6 +168,24 @@ func (ni *NodeInfo) setNodeState(node *v1.Node) {
...
 }
 
+func (ni *NodeInfo) SetNodeGPUInfo(node *v1.Node) {
+
+	core, ok := node.Status.Capacity["volcano.sh/node-gpu-core"]
+	if ok {
+		ni.GPUTotalCore = int(core.Value())
+	}
+
+	mem, ok := node.Status.Capacity["volcano.sh/node-gpu-memory"]
```
Comment (hzxuzhonghu) [inline pkg/scheduler/api/node_info.go:178]: Let's declare with a const, same applies to "volcano.sh/node-gpu-core"

File: pkg/scheduler/api/node_info.go [hunk 5]
```diff
@@ -46,8 +46,7 @@ type NodeInfo struct {
 
 	Tasks map[TaskID]*TaskInfo
 
-	// Used to store custom information
-	Others map[string]interface{}
+	GPUDevices map[int]*GPUDevice
```
Comment (k82cn) [inline pkg/scheduler/api/node_info.go:49]: please do not remove `Others`, some projects are using it.

Comment (hzxuzhonghu) [reply pkg/scheduler/api/node_info.go:51]: reverted

File: pkg/scheduler/api/node_info.go [hunk 6]
```diff
@@ -154,9 +145,35 @@ func (ni *NodeInfo) setNodeState(node *v1.Node) {
...
+	res, ok := node.Status.Capacity[VolcanoGPUNumber]
+	if !ok {
+		return
+	}
+	gpuNumber := res.Value()
+	if gpuNumber == 0 {
+		klog.Warningf("invalid %s=%s", VolcanoGPUNumber, res.String())
+	}
+
+	memoryPerCard := uint(totalMemory / gpuNumber)
```
Comment (carmark) [inline pkg/scheduler/api/node_info.go:167]: gpuNumber may be `0` ?

Comment (hzxuzhonghu) [reply pkg/scheduler/api/node_info.go:167]: Correct, need to fix

Comment (hzxuzhonghu) [reply pkg/scheduler/api/node_info.go:168]: done

**Pull request merged**